### PR TITLE
WIP calls

### PR DESCRIPTION
**Pull request by bnjbvr**

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 1]
```diff
@@ -1,37 +1,59 @@
...
+    Reg(RealReg, ir::Type),
+    Stack(i64, ir::Type),
 }
 
-#[derive(Clone, Debug)]
-enum ABIRet {
-    Reg(RealReg),
-    _Stack,
+/// X64 ABI information shared between body (callee) and caller.
+struct ABISig {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs:28]: General note on the ABI code: we should try to factor out the commonalities between `x64/abi.rs` and `aarch64/abi.rs`. (I know it was already like this, except with more divergence, so I'm not faulting this patch at all!) I'm thinking it might be best to actually do this as part of the merge of this patch, but if that turns out to be too much work, then we can defer a bit.

Related to that, while working on reftypes it seemed to me that an arch-independent "stack frame layout" descriptor wants to escape from all of this and live in the compilation result. I might refactor in that direction later. Or perhaps we could unify x64 and aarch64 at that point.

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 2]
```diff
@@ -298,36 +274,75 @@ impl ABIBody for X64ABIBody {
...
+                let from_bits = ty_bits(ty) as u8;
+                let ext_mode = match from_bits {
+                    1 | 8 => Some(ExtMode::BQ),
+                    16 => Some(ExtMode::WQ),
+                    32 => Some(ExtMode::LQ),
+                    64 => None,
+                    _ => unreachable!(),
+                };
+
+                // Trash the from_reg; it should be its last use. XXX is this safe to do so?
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs]: Yep, should be safe, because `from_reg` is passed in as a `Writable<Reg>`. (I think the PR that changed `from_reg` to make this safer came after `abi.rs` was forked for x64, so this might have been a bit confusing; sorry!) We can remove the conversion from writable-reg to reg to writable-reg on the next line as well to make this more clear.

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 3]
```diff
@@ -298,36 +274,75 @@ impl ABIBody for X64ABIBody {
...
+                        ret.push(Inst::movsx_rm_r(
+                            ext_mode,
+                            RegMem::reg(from_reg.to_reg()),
+                            dest_reg,
+                        ));
+                    }
+                    _ => {}
+                };
+
+                debug_assert!(
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs]: I'd prefer `assert!` rather than `debug_assert!` for offset limit checks (even if we'll eventually fix this; just to make sure we don't commit something that skips a security check).

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 4]
```diff
@@ -393,20 +410,14 @@ impl ABIBody for X64ABIBody {
...
-                _ => unimplemented!(),
-            }
-        }
+        let callee_saved_used = clobbered
+            .iter()
+            .map(|reg| match reg.to_reg().get_class() {
+                RegClass::I64 => 8,
+                _ => todo!(),
+            })
+            .fold(0, |a, b| a + b);
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs]: `.sum()` ?

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 5]
```diff
@@ -441,6 +452,25 @@ impl ABIBody for X64ABIBody {
...
+                RegClass::I64 => {
+                    insts.push(Inst::push64(RegMemImm::reg(r_reg.to_reg())));
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        if callee_saved_used > 0 {
+            insts.push(Inst::VirtualSPOffsetAdj {
+                offset: -(callee_saved_used as i64),
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs]: This should be a positive offset, I think -- if nominal-sp was `[esp+N]` before we push, it's `[esp+N+callee_saved]` afterward.

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 6]
```diff
@@ -453,13 +483,34 @@ impl ABIBody for X64ABIBody {
...
+                        w_real_reg.to_reg().to_reg(),
+                    )));
+                    callee_saved_used += 8;
+                }
+                _ => unimplemented!(),
+            }
+        }
+
+        if callee_saved_used > 0 {
+            insts.push(Inst::VirtualSPOffsetAdj {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/abi.rs]: It's a little subtle, but we actually *don't* want to do this here, because of how mid-function returns interact with the virtual-SP-offset tracking during emission. (Basically, we don't track virtual-SP-offset along the CFG edges, but only in a straight line, so a mid-function return epilogue would throw off everything below it.) It's OK not to adjust because the epilogue is the last code that runs before return, and so we can just be sure to not use nominal-SP refs.

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 7]
```diff
@@ -960,38 +972,41 @@ pub(crate) fn emit(inst: &Inst, sink: &mut MachBuffer<Inst>) {
...
-            // adjustment.
-            let taken_disp = taken.as_offset32_or_zero() - 6;
-            let taken_disp = taken_disp as u32;
+            // If taken.
             let cond_start = sink.cur_offset();
             let cond_disp_off = cond_start + 2;
             let cond_end = cond_start + 6;
             if let Some(l) = taken.as_label() {
                 sink.use_label_at_offset(cond_disp_off, l, LabelUse::Rel32);
+                // In the payload field, encoded a negative addend of -4 (0xFFFFFFFC, reversed
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs]: nit: s/encoded/encode/

File: cranelift/codegen/src/isa/x64/inst/emit.rs [hunk 8]
```diff
@@ -960,38 +972,41 @@ pub(crate) fn emit(inst: &Inst, sink: &mut MachBuffer<Inst>) {
             taken,
             not_taken,
         } => {
-            // Conditional part.
+            // The Intel encoding requires an offset from the beginning of the next instruction to
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/emit.rs]: I wonder if it might be cleaner to incorporate this into the `LabelUse` implementation for x64? E.g., we have a particular type of PC-rel label reference that resolves relative to the end of the label, not the beginning.

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 9]
```diff
@@ -300,8 +341,8 @@ impl Inst {
         Inst::EpiloguePlaceholder
     }
 
-    pub(crate) fn jmp_known(dest: BranchTarget) -> Inst {
-        Inst::JmpKnown { dest }
+    pub(crate) fn jmp_known(dst: BranchTarget) -> Inst {
+        Inst::JmpKnown { dst }
     }
 
     pub(crate) fn jmp_cond_symm(cc: CC, taken: BranchTarget, not_taken: BranchTarget) -> Inst {
```
Comment (cfallin) [inline cranelift/codegen/src/isa/x64/inst/mod.rs]: Wasn't changed in this PR, but a drive-by comment: it might be nice to rename `JmpCondSymm` to just `JmpCond`, now that we don't have the notion of symmetric (2-target) vs. 1-target forms and branch resolution at this layer anymore.

**Review by cfallin — APPROVED** [review]
This is fantastic work! As noted in the comments below, I definitely want to look for ways to factor out common bits; but having a solid x64 implementation first is an excellent starting point for that :-)

I didn't think as much about the design of `x64::Inst` and the arguments it takes; I'm not sure if @julian-seward1 might have more thoughts on that. In any case, I'm happy with it as-is.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 1 above)
Comment (cfallin) [reply cranelift/codegen/src/isa/x64/abi.rs:28]: (Addendum after sending review: probably best to do any refactoring after an initial merge; so ignore that part of above.)

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 10]
```diff
@@ -264,27 +242,25 @@ impl ABIBody for X64ABIBody {
...
-                    return Inst::mov_r_r(/*is64=*/ true, from_reg.to_reg(), to_reg);
-                } else if from_reg.get_class() == RegClass::V128 {
-                    // TODO: How to support Movss. Should is64 always be true?
-                    return Inst::xmm_r_r(SseOpcode::Movsd, from_reg.to_reg(), to_reg);
-                }
-                unimplemented!("moving from non-int arg to vreg {:?}", from_reg.get_class());
+        match &self.sig.args[idx] {
+            ABIArg::Reg(from_reg, ty) => Inst::gen_move(to_reg, from_reg.to_reg(), *ty),
+            &ABIArg::Stack(off, ty) => {
+                debug_assert!(self.fp_to_arg_offset() + off <= u32::max_value() as i64,);
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/abi.rs]: Is it right to compare against `u32::max_value()` here?  IIUC x64 interprets `reg + imm32` addresses as `reg +64 sx_32_to_64(imm32)`.  Hence if `elf.fp_to_arg_offset() + off` exceeds `i32::max_value` it will wrap around and be considered a large negative value.   (Also applies to various other assertions further down)

Also .. it seems to me that what this assertion guards against isn't bugs in our code, but rather an unhandled -- if very unlikely -- case of more than 2^31 bytes in the stack frame.  Or 2^31 bytes of args.  Or some such.  Could/should we handle such cases more cleanly?

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 11]
```diff
@@ -520,3 +554,482 @@ impl ABIBody for X64ABIBody {
...
+    caller_saved.push(Writable::from_reg(regs::rdi()));
+    caller_saved.push(Writable::from_reg(regs::rax()));
+    caller_saved.push(Writable::from_reg(regs::rcx()));
+    caller_saved.push(Writable::from_reg(regs::rdx()));
+    caller_saved.push(Writable::from_reg(regs::r8()));
+    caller_saved.push(Writable::from_reg(regs::r9()));
+    caller_saved.push(Writable::from_reg(regs::r10()));
+    caller_saved.push(Writable::from_reg(regs::r11()));
+
+    // - FPU: all the registers!
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/abi.rs]: As a nano-nit, "FPU" on Intels is sometimes interpreted as "the legacy x87 FP facilities".  You really mean "XMM state" here -- the FPU state is ST0 .. ST7.

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 12]
```diff
@@ -3,16 +3,19 @@
...
 use crate::machinst::*;
 
-use super::regs::show_ireg_sized;
+use super::{
+    regs::{self, show_ireg_sized},
+    EmitState,
+};
 
-/// A Memory Address. These denote a 64-bit value only.
+/// A real Memory Address. These denote a 64-bit value only.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/args.rs]: nano-nit: What does "real" mean here?  This is actually a virtual address ..

File: cranelift/codegen/src/isa/x64/inst/args.rs [hunk 13]
```diff
@@ -84,14 +82,81 @@ impl ShowWithRRU for Addr {
     }
 }
 
+/// A Memory Address. These denote a 64-bit value only.
+/// Used for real addressing modes as well as addressing modes used during compilation, when the
+/// moving SP offset is not known.
+#[derive(Clone)]
+pub enum VirtualAddr {
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/args.rs]: I kind-of see why you changed `Addr` to `VirtualAddr`.  A couple of comments .. firstly, the `Virtual` bit doesn't convey much new info, since it's impossible to operate the machine (at least in user mode, and I suspect also in kernel mode) without having the MMU running; so all addresses are virtual addresses, from our perspective.  Secondly, technically these are not addresses; what they are are expressions (that are part of instructions), which when evaluated at run time, produce addresses.  Ancient Intel documentation (I have a 486 manual!) calls these "Address Modes", and more recent docs appear to use the term "effective address", although somewhat ambiguously.  Hennessy & Patterson appears to use the term "addressing mode".

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 14]
```diff
@@ -100,46 +102,71 @@ pub(crate) enum Inst {
...
+    // =====================================
+    // Floating-point operations.
+    /// Float arithmetic/bit-twiddling: (add sub and or xor mul adc? sbb?) (32 64) (reg addr imm) reg
+    XMM_RM_R {
+        op: SseOpcode,
+        src: RegMem,
+        dst: Writable<Reg>,
+    },
+
+    /// Float mvoes: mov (64 32) reg reg
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs]: (1) typo `mvoes`, (2) since `op` is `SseOpcode`, can't this be move than just moves?  In which case doesn't `XMM_RM_R` make it redundant?

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 15]
```diff
@@ -148,22 +175,20 @@ pub(crate) enum Inst {
...
-        src: Reg,
-        dst: Writable<Reg>,
-    },
+    /// An instruction that will always trigger the illegal instruction exception.
+    Ud2 { trap_info: (SourceLoc, TrapCode) },
+
+    // =====================================
+    // Meta-instructions generating no code.
+    /// Marker, no-op in generated code: SP "virtual offset" is adjusted. This
+    /// controls MemArg::NominalSPOffset args are lowered.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs]: typo: += `how` ?

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 16]
```diff
@@ -215,6 +359,185 @@ fn lower_insn_to_regs<'a>(ctx: Ctx<'a>, inst: IRInst) {
...
+                Opcode::Load
+                | Opcode::Uload8
+                | Opcode::Sload8
+                | Opcode::Uload16
+                | Opcode::Sload16
+                | Opcode::Uload32
+                | Opcode::Sload32 => {
+                    assert!(inputs.len() == 1, "only one input for load operands");
+                    let base = input_to_reg(ctx, inputs[0]);
+                    // TODO is this the right thing for the offset's sign?
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs]: I don't know; but what I can say is that the machine will interpret the offset as `i32` not `u32` (even though it's stored as `u32`; that's OK).

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 17]
```diff
@@ -215,6 +359,185 @@ fn lower_insn_to_regs<'a>(ctx: Ctx<'a>, inst: IRInst) {
...
+                    Addr::imm_reg_reg_shift(offset as u32, base, index, shift)
+                }
+
+                _ => unreachable!(),
+            };
+
+            let dst = output_to_reg(ctx, outputs[0]);
+            match (sign_extend, is_float) {
+                (true, false) => {
+                    // Sign-extended moves only happen for smaller-than-machine-word size.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/lower.rs]: Can you clarify that comment?  I'm not sure what it means.  For the destination being smaller than the machine word?  But what about `movs{b,w,l}q` then?

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 1 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs:28]: Agreed that since the common API is stil a moving target until the full x64 is implemented, it might be better to do it after the impl work. (Fwiw, I'd be happy to do these refactorings.)

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 3 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs]: Good call! All of these asserts were put here to make sure that we'd run into them once they're needed; i expect the realistic wasm programs will run into this.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 4 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs]: TIL, thanks.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 6 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs]: Noted; will comment.

File: cranelift/codegen/src/isa/x64/inst/emit.rs (hunk 8 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/emit.rs]: Good idea, done.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 10 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs]: You are totally right for both points; it happens so that Rust in debug mode will panic if there's an int overflow on add, and that I've used most of these assertions as spots to fix later. I'll do it quickly after landing this!

In the meanwhile, I'll change this to an assertion.

File: cranelift/codegen/src/isa/x64/inst/args.rs (hunk 13 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/args.rs]: Settled to Amode and SyntheticAmode, as discussed on zulip.

File: cranelift/codegen/src/isa/x64/inst/mod.rs (hunk 14 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/inst/mod.rs]: I *think* that Johnnie's PR takes care of the redundancy, or else I'll add it to my todo list, closing at the moment.

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 16 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/lower.rs]: Ah thanks, this was what I was wondering about.

File: cranelift/codegen/src/isa/x64/lower.rs (hunk 17 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/lower.rs]: I've replaced it by the following:

```
                    // The load is sign-extended only when the output size is lower than 64 bits,
                    // so ext-mode is defined in this case.
```

**Comment by bnjbvr** [comment]
Will reopen against upstream.
